### DISCUSSION TITLE
正确更新铁蛋内部Jetpack版本的方法

### DISCUSSION THREAD

**Discussion by homalozoa** [MiRoboticsLab/cyberdog_ros2#110] [General]
## 引

有朋友在https://github.com/MiRoboticsLab/cyberdog_ros2/issues/63 提出如何更新铁蛋内部的Jetpack，为了方便讨论，我直接挪到了Discussion里：https://github.com/MiRoboticsLab/cyberdog_ros2/discussions/83

导致该问题的原因是，在我们研发铁蛋过程中修改了部分Tegra内核的内容，并且替换了维护原内核的deb包。也就是`nvidia-l4t-kernel`相关的一些包。这一问题其实可以使用[dkms](https://github.com/dell/dkms)技术解决，但限于目前精力有限，姑且采取本文所述的临时方案解决此问题。

## 更新原理

本文所述的更新方法是，首先使用`apt-mark hold`锁住当前系统中的一些`内核`相关的软件包，而后修改源的版本号，再做相应的安装和更新。最后，再做一些善后工作，即铁蛋的一些配置项。即可完成更新。

### 确认版本号

```
cat /etc/nv_tegra_release
```

此时应该返回

```
# R32 (release), REVISION: 5.0, GCID: 25531747, BOARD: t186ref, EABI: aarch64, DATE: Fri Jan 15 23:21:05 UTC 2021
```

### 锁住内核相关软件包版本

```
sudo apt-mark hold nvidia-l4t-kernel nvidia-l4t-kernel-dtbs nvidia-l4t-kernel-headers
```

### 修改源版本号

```
sudo vim /etc/apt/sources.list.d/nvidia-l4t-apt-source.list
```

修改其中 `32.5` 到 `32.6`。

### 整体更新

```
sudo apt update
sudo apt install nvidia-jetpack
sudo apt upgrade
```

*期间可能会因为网络问题导致失败，重试即可*

*期间会提示是否替换一些配置文件，按Y即可*

### 善后工作

首先通过指令确认是否更新完毕

```
cat /etc/nv_tegra_release
```

看版本号是否和下述内容一致。

```
# R32 (release), REVISION: 6.1, GCID: 27863751, BOARD: t186ref, EABI: aarch64, DATE: Mon Jul 26 19:36:31 UTC 2021
```

最后，使能网卡内部广播后，重启即可。

```
sudo systemctl enable cyberdog_multicast.service
sudo reboot
```

**Comment by Oldpan** [comment]
感谢海总~这个更新后，确认cuda可以正常工作吗？

**Reply by homalozoa** [reply]
貌似还有点问题，需要在kernel打个补丁。32.6对这部分有修改。

**Reply by Oldpan** [reply]
这个问题能解决么....如果cuda不能用，其实更新的意义就没啥了

**Reply by homalozoa** [reply]
一定可以的hhh，我是没考虑到更新之后cuda会有异常，我最近再研究一下

**Reply by Oldpan** [reply]
嗯嗯，辛苦了~  我对kernel底层不是很熟悉，就靠海果大大了

**Comment by zbwu** [comment]
用dkms估计也不好解决，因为还涉及DTS修改。如果像Raspberry Pi那样用overlay是灵活，但维护成本更高，Cyberdog是一个固定的产品，也不需要这种灵活性。还是独立维护一个和nvidia官方同步的kernel比较好。

**Reply by homalozoa** [reply]
确实，维护一个独立的才好解决这个问题
